src/cgutils.cpp

```diff
@@ -1060,13 +1060,18 @@ static void emit_type_error(jl_codectx_t &ctx, const jl_cgval_t &x, Value *type,
 
 static std::pair<Value*, bool> emit_isa(jl_codectx_t &ctx, const jl_cgval_t &x, jl_value_t *type, const std::string *msg)
 {
+    // TODO: The subtype check below suffers from incorrectness issues due to broken
+    // subtyping for kind types (see https://github.com/JuliaLang/julia/issues/27078). For
+    // actual `isa` calls, this optimization should already have been performed upstream
+    // anyway, but having this optimization in codegen might still be beneficial for
+    // `typeassert`s if we can make it correct.
     Optional<bool> known_isa;
     jl_value_t *intersected_type = type;
     if (x.constant)
         known_isa = jl_isa(x.constant, type);
-    else if (jl_subtype(x.typ, type))
-        known_isa = true;
-    else {
+    else if (jl_subtype(x.typ, type)) {
+        // known_isa = true;
+    } else {
```

Comment:
Okay, the minimal change that my tests hinge on seems to be here. [Here is a gist of the tests I'm using to check this PR](https://gist.github.com/jrevels/4e56e78d91d53ca29af73b5b36d506d0).

Before commenting out `known_isa = true`, some of the tests fail. 

After commenting out `known_isa = true`, all of them pass except for the `g(identity, 1)` MWE reported earlier.

Does anybody know (or can help figure out) how to get all of these tests passing so this can be merged?